src/SourceBrowser.Site/Controllers/UploadController.cs

```diff
@@ -31,64 +31,86 @@ public ActionResult Submit(string githubUrl)
-                htmlTransformer.Visit(workspaceModel);
+                var organizationPath = System.Web.Hosting.HostingEnvironment.MapPath("~/") + "SB_Files\\" + retriever.UserName;
+                var repoPath = Path.Combine(organizationPath, retriever.RepoName);
 
-                var searchTransformer = new SearchIndexTransformer(retriever.UserName, retriever.RepoName);
-                searchTransformer.Visit(workspaceModel);
+                // TODO: Use parallel for.
+                // TODO: Process all solutions.
+                // For now, we're assuming the shallowest and shortest .sln file is the one we're interested in
+                foreach (var solutionPath in solutionPaths.OrderBy(n => n.Length).Take(1))
+                {
+                    Generator.Model.WorkspaceModel workspaceModel;
+                    try
+                    {
+                        workspaceModel = UploadRepository.ProcessSolution(solutionPath, repoRootPath);
+                    }
+                    catch (Exception ex)
+                    {
+                        ViewBag.Error = "There was an error processing solution " + Path.GetFileName(solutionPath);
+                        return View("Index");
+                    }
 
+                    //One pass to lookup all declarations
+                    var typeTransformer = new TokenLookupTransformer();
+                    typeTransformer.Visit(workspaceModel);
+                    var tokenLookup = typeTransformer.TokenLookup;
 
+                    //Another pass to generate HTMLs
+                    var htmlTransformer = new HtmlTransformer(tokenLookup, repoPath);
+                    htmlTransformer.Visit(workspaceModel);
 
-                // Generate HTML of the tree view
-                var treeViewTransformer = new TreeViewTransformer(repoPath, retriever.UserName, retriever.RepoName);
-                treeViewTransformer.Visit(workspaceModel);
-            }
+                    var searchTransformer = new SearchIndexTransformer(retriever.UserName, retriever.RepoName);
+                    searchTransformer.Visit(workspaceModel);
 
-            return Redirect("/Browse/" + retriever.UserName + "/" + retriever.RepoName);
+                    // Generate HTML of the tree view
+                    var treeViewTransformer = new TreeViewTransformer(repoPath, retriever.UserName, retriever.RepoName);
+                    treeViewTransformer.Visit(workspaceModel);
+                }
+
+                processingSuccessful = true;
+                return Redirect("/Browse/" + retriever.UserName + "/" + retriever.RepoName);
+            }
+            finally
+            {
+                if (processingSuccessful)
+                {
+                    BrowserRepository.MarkRepositoryReady(retriever.UserName, retriever.RepoName);
+                }
+                else
+                {
+                    BrowserRepository.UnlockRepository(retriever.UserName, retriever.RepoName);
```

Comment:
Sure. Here is what happens in the first scenario:

User 1: Uploading a repo.
`Uploading. Please wait...`
User 2: Trying to upload the same repo (the state is PROCESSING.)
`We are currently processing AmadeusW's repository ece327-lab3-helper.
This may take a few minutes. Refresh this page periodically.`
User 1: The upload finished
`Browsing AmadeusW's repository ece327-lab3-helper.
Use the search box or the tree view on the right to browse within this repository.`
User 3: Trying to upload the same repo (the state is READY.)
`Browsing AmadeusW's repository ece327-lab3-helper.
Use the search box or the tree view on the right to browse within this repository.`


Thanks for bringing up the second scenario! I made a bug. :bug:
That's how the code will execute now:

User 1: Uploading a repo
`Uploading. Please wait...`
User 2 is told to wait
`We are currently processing AmadeusW's repository ece327-lab3-helper.
This may take a few minutes. Refresh this page periodically.`
User 1 is notified of an exception during upload
`Upload any Github Repo
Paste the URL of the project
There was an error processing solution ece327lab3helper.sln`
User 2: Refreshing the page
`There was and error retrieving this project
Please ensure the URL you're using is correct.
An error occured during processing of repository ece327-lab3-helper`
This is because `if(!BrowserRepository.PathExists(username, repository))` (BrowseController.cs:38) returns *true* and `if (!BrowserRepository.RepositoryIsReady(username, repository))` (BrowseController.cs:47) throws an exception.
User is able to re-upload the repo (because *repo.lock* has been deleted)